tls/s2n_prf.c

```diff
@@ -326,35 +330,67 @@ const struct s2n_p_hash_hmac *s2n_get_hmac_implementation() {
 #endif
 }
 
-int s2n_prf_new(struct s2n_connection *conn)
+S2N_RESULT s2n_prf_new(struct s2n_connection *conn)
 {
-    /* Set p_hash_hmac_impl on initial prf creation. 
-     * When in FIPS mode, the EVP API's must be used for the p_hash HMAC.
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_EQ(conn->prf_space, NULL);
+
+    DEFER_CLEANUP(struct s2n_blob data = { 0 }, s2n_free);
+    RESULT_GUARD_POSIX(s2n_realloc(&data, sizeof(struct s2n_prf_working_space)));
+    RESULT_GUARD_POSIX(s2n_blob_zero(&data));
+    conn->prf_space = (struct s2n_prf_working_space*)(void*) data.data;
+    ZERO_TO_DISABLE_DEFER_CLEANUP(data);
+
+    conn->prf_space->p_hash_hmac_impl = s2n_get_hmac_implementation();
+    RESULT_GUARD_POSIX(conn->prf_space->p_hash_hmac_impl->alloc(conn->prf_space));
```

Comment:
I would love a comment here explaining why we are calling an alloc function on a pointer that we just allocated memory to.